### PR TITLE
gateway/azure: Parse error responses for anonymous requests

### LINKED ISSUE #4481: [gateway-azure] curl to a bucket causes 500 internal error
## Expected Behavior

curl should not cause 500 errors in the server.

## Current Behavior

curl causes 500 errors when accessing many urls in Azure gateway mode.

## Possible Solution

Improve validation.

## Steps to Reproduce (for bugs)

1. Start Minio in Azure gateway mode.
2. Use to make any request to a "bucket" url:
```
$ curl -v -s "http://localhost:9000/asdfasdf"
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 9000 (#0)
> GET /asdfasdf HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Accept-Ranges: bytes
< Content-Type: application/xml
< Server: Minio/RELEASE.2017-06-02T22-13-34Z (linux; amd64)
< Vary: Origin
< X-Amz-Bucket-Region: 
< X-Amz-Request-Id: 14C56205E722FB40
< Date: Tue, 06 Jun 2017 00:50:28 GMT
< Transfer-Encoding: chunked
< 
<?xml version="1.0" encoding="UTF-8"?>
* Connection #0 to host localhost left intact
<Error><Code>InternalError</Code><Message>We encountered an internal error, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/asdfasdf</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
```
On server, we see an error log like:
```
ERRO[0085] Unable to list objects.                       cause=expected element type <EnumerationResults> but have <Error> source=[gateway-handlers.go:767:gatewayAPIHandlers.ListObjectsV1Handler()] stack=home/builder/mygo/src/github.com/minio/minio/cmd/gateway-azure-anonymous.go:170:(*azureObjects).AnonListObjects home/builder/mygo/src/github.com/minio/minio/cmd/gateway-handlers.go:760:(GatewayLayer).AnonListObjects-fm home/builder/mygo/src/github.com/minio/minio/cmd/gateway-handlers.go:765:gatewayAPIHandlers.ListObjectsV1Handler home/builder/mygo/src/github.com/minio/minio/cmd/gateway-router.go:105:(gatewayAPIHandlers).ListObjectsV1Handler-fm
```
3. On plain old Minio server this is perfectly handled:
```
$ curl -v -s "http://localhost:9000/asdfasdf"
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 9000 (#0)
> GET /asdfasdf HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Accept-Ranges: bytes
< Content-Type: application/xml
< Server: Minio/DEVELOPMENT.2017-06-05T18-57-45Z (linux; amd64)
< Vary: Origin
< X-Amz-Bucket-Region: us-east-1
< X-Amz-Request-Id: 14C562C6A068935A
< Date: Tue, 06 Jun 2017 01:04:16 GMT
< Transfer-Encoding: chunked
< 
<?xml version="1.0" encoding="UTF-8"?>
* Connection #0 to host localhost left intact
<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Message><Key></Key><BucketName></BucketName><Resource>/asdfasdf</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
```
4.

## Your Environment

* Version used (`minio version`):
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version (`uname -a`):
* Link to your project:

### PR DESCRIPTION
**Pull request by krishnasrinivas**
fixes #4481

## Description

For anonymous requests to Azure we were not parsing the Error XML responses.

## How Has This Been Tested?

manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by harshavardhana** [comment]
@krishnasrinivas can you add unit tests for the changes?

**Comment by krishnasrinivas** [comment]
@harshavardhana azureAnonRequest can't be unit tested as it needs to talk to azure backend. This would be covered under mint.

**Comment by harshavardhana** [comment]
> @harshavardhana azureAnonRequest can't be unit tested as it needs to talk to azure backend. This would be covered under mint.

The parsing part if possible can go in unit test since it is specific for azure to s3 translation.

**Comment by krishnasrinivas** [comment]
> The parsing part if possible can go in unit test since it is specific for azure to s3 translation.

```
+		if len(respBody) == 0 {
 +			// no error in response body, might happen in HEAD requests
 +			return nil, storage.AzureStorageServiceError{
 +				StatusCode: resp.StatusCode,
 +				Code:       resp.Status,
 +				Message:    "no response body was available for error status code",
 +			}
 +		}
 +		// Response contains Azure storage service error object.
 +		var storageErr storage.AzureStorageServiceError
 +		if err := xml.Unmarshal(respBody, &storageErr); err != nil {
 +			return nil, err
 +		}
 +		storageErr.StatusCode = resp.StatusCode
 +		return nil, storageErr
```

This part can be separated out into a function, but the downside is it will create a small function which is not reused. Let me check with @abperiasamy once what he thinks

**Comment by harshavardhana** [comment]
> This part can be separated out into a function, but the downside is it will create a small function which is not reused. Let me check with @abperiasamy once what he thinks

It doesn't need to be reused the problem is that we have many functions like this which are not unit tested because they are deep inside, if we do have one function for this we can add test cases for all common and corner cases where this can fail. 

Say if we do reflection or something internally.  My suggestion is we add tests..

File: cmd/gateway-azure-anonymous.go [hunk 1]
```diff
@@ -29,18 +29,59 @@ import (
...
+	}
+	if header != nil {
+		req.Header = header
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode >= 400 && resp.StatusCode <= 505 {
```
Comment (krisis) [inline cmd/gateway-azure-anonymous.go]: A comment explaining why we pick status code between 400 and 505 is required.

**Review by krisis — CHANGES_REQUESTED** [review]
Overall the changes look good to me. @krishnasrinivas do you plan to add unit tests? I see that there was a discussion but no decision mentioned in the conversation.

File: cmd/gateway-azure-anonymous.go [hunk 2]
```diff
@@ -29,18 +29,59 @@ import (
 	"github.com/Azure/azure-sdk-for-go/storage"
 )
 
+// Make anonymous HTTP request to azure endpoint.
+func azureAnonRequest(verb, urlStr string, header http.Header) (*http.Response, error) {
```
Comment (balamurugana) [inline cmd/gateway-azure-anonymous.go:33]: Please add unit test to this function.  You could run a test http server in the unit test and address various cases.

File: cmd/gateway-azure-anonymous.go [hunk 3]
```diff
@@ -29,18 +29,59 @@ import (
...
+		if err != nil {
+			return nil, err
+		}
+
+		if len(respBody) == 0 {
+			// no error in response body, might happen in HEAD requests
+			return nil, storage.AzureStorageServiceError{
+				StatusCode: resp.StatusCode,
+				Code:       resp.Status,
+				Message:    "no response body was available for error status code",
```
Comment (balamurugana) [inline cmd/gateway-azure-anonymous.go:59]: Why is missing response body wrong?

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: cmd/gateway-azure-anonymous.go (hunk 2 above)
Comment (krishnasrinivas) [reply cmd/gateway-azure-anonymous.go:33]: We had decided not to do mock tests. i.e we should not start a mock server and test instead should use the actual backend in functional tests in mint to cover cases like these.

File: cmd/gateway-azure-anonymous.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/gateway-azure-anonymous.go]: I just picked this code from azure sdk
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes says 4xx 5xx are error codes

File: cmd/gateway-azure-anonymous.go (hunk 3 above)
Comment (krishnasrinivas) [reply cmd/gateway-azure-anonymous.go:59]: @balamurugana note that this is inside the if condition: 
```
if resp.StatusCode >= 400 && resp.StatusCode <= 505
```

hence it is an error condition regardless of whether body is available or not

**Comment by krishnasrinivas** [comment]
> Overall the changes look good to me. @krishnasrinivas do you plan to add unit tests? I see that there was a discussion but no decision mentioned in the conversation.

@krisis AB said functions like this which needs to talk to GCS backend should be covered under functional tests (to test against the actual backend) and that we should not do mock/unit tests.

**Review by balamurugana — APPROVED** [review]
Not tested

**Review by krisis — APPROVED** [review]
Looks like we have decided what to do about unit tests. Lgtm otherwise.
